infrastructure/smart-contracts/contracts/Achievements.sol

```diff
@@ -0,0 +1,147 @@
+
+    //state
+    IAccessControl public factoryAddress; //factory proxy to check access control
+    mapping(uint256 => bool) public definedTokens; //tokenId => exists
+    uint256[] private _tokenIdList; //array of token ids that exist (ordered by mint order)
+    uint256 public tokenIdCount; //counter for unique token ids that exist
+    uint256 public totalSupply; //counter for total tokens minted on this contract
+    mapping(uint256 => uint256) public totalSupplyById; //tokenId => tokenCount
+
+    /**
+     * @dev Achievements constructor.
+     * @param _factoryAddress Address of the factory contract that produced this contract.
+     * @param _uri URI string pointing to metadata (e.g. https://metadata.xai.games/id.json)
+     */
+    constructor(address _factoryAddress, string memory _uri) ERC1155(_uri) {
+        factoryAddress = IAccessControl(_factoryAddress);
+    }
+
+    /**
+     * @dev Restricts caller to MINT_ROLE holder in factory contract.
+     */
+    modifier onlyMintRole() {
+        require(factoryAddress.hasRole(MINT_ROLE, address(msg.sender)), "caller does not have MINT_ROLE");
+        _;
+    }
+
+    /**
+     * @dev Mints a single token to the recipient address. Only callable by MINT_ROLE holder.
+     * @param to Address to receive the newly minted token.
+     * @param id Token ID of token to mint.
+     * @param data Arbitrary data field.
+     */
+    function mint(address to, uint256 id, bytes memory data) public onlyMintRole {
+        require(balanceOf(to, id) == 0, "address has non-zero token balance");
+        
+        if (!definedTokens[id]) {
+            _tokenIdList.push(id);
+            tokenIdCount += 1;
+        }
+        totalSupply += 1;
+        totalSupplyById[id] += 1;
+
+        _mint(to, id, 1, data);
+    }
+
+    /**
+     * @dev Mints a batch of tokens to the recipient address. Only callable by MINT_ROLE holder.
+     * @param to Address to receive the newly minted tokens.
+     * @param ids Array of token IDs of tokens to mint.
+     * @param data Arbitrary data field.
+     */
+    function mintBatch(address to, uint256[] calldata ids, bytes memory data) public onlyMintRole {
+        require(ids.length > 0, "invalid ids param length");
+        
+        uint256[] memory amounts = new uint256[](ids.length);
+        for (uint256 i = 0; i < ids.length; i++) {
+            require(balanceOf(to, ids[i]) == 0, "address has non-zero token balance");
+            
+            if (!definedTokens[ids[i]]) {
+                _tokenIdList.push(ids[i]);
+                tokenIdCount += 1;
+            }
+            totalSupply += 1;
+            totalSupplyById[ids[i]] += 1;
+            amounts[i] = 1;
+        }
+
+        _mintBatch(to, ids, amounts, data);
+    }
+
+    /**
+     * @dev Gets an array of all defined token ids for this contract.
+     */
+    function getDefinedTokens() external view returns (uint256[] memory) {
+        return _tokenIdList;
+    }
```

Comment:
If we make `_tokenIdList` public this function could be removed and we save some space.